### PR TITLE
microRTPS: fix UART link

### PR DESCRIPTION
**Pull request by TSC21**
Currently the microRTPS bridge UART link is broken and it's resulting on continuous CRC errors on the agent side that after some seconds close the link. This is being investigated.

Also, it was verified that a stack increase was required to avoid the `micrortps_client` of causing and hardfault and rebooting the flight controller.

### CODE REVIEW

**Comment by TSC21** [comment]
@AndrewWedemier thanks for your insights and your debugging effort, which ended-up helping me unblock this! Added you as a co-author in one commit, which was the commit that allows to have the bridge usable over UART.

Note that for setups which are to run at baud rates <=1MByte/sec, it's required to throttle the overall publishing using the `-s` option. So with the current default streams (per 07/20/2020), something like:

```sh
micrortps_client start -d /dev/ttyS2 -b 1000000 -v -w 6
./micrortps_agent -d /dev/ttyUSB1 -b 1000000 -v -w 6
```
works nice without any errors, though it introduces some latency on the link. The alternative is also increasing the baud rate. Found that using 1.5MByte/s with the current default streams is enough to have it work properly.

```sh
micrortps_client start -d /dev/ttyS2 -b 1500000 -v
./micrortps_agent -d /dev/ttyUSB1 -b 1500000 -v
```

Note that with the increase of the number of topics being streamed, and depending on each rates, this might change and might require changes on the stream rates.

File: msg/templates/urtps/microRTPS_transport.cpp [hunk 1]
```diff
@@ -113,104 +113,101 @@ ssize_t Transport_node::read(uint8_t *topic_ID, char out_buffer[], size_t buffer
 
 	*topic_ID = 255;
 
-	ssize_t len = node_read((void *)(rx_buffer + rx_buff_pos), sizeof(rx_buffer) - rx_buff_pos);
+	ssize_t len = 0;
 
-	if (len <= 0) {
-		int errsv = errno;
+	while (0 < (len = node_read((void *)(rx_buffer + rx_buff_pos), sizeof(rx_buffer) - rx_buff_pos))) {
```
Comment (bkueng) [inline msg/templates/urtps/microRTPS_transport.cpp:118]: Why was this changed?

File: msg/templates/urtps/microRTPS_transport.cpp [hunk 2]
```diff
@@ -225,13 +222,12 @@ ssize_t Transport_node::write(const uint8_t topic_ID, char buffer[], size_t leng
...
 
 	static struct Header header = {{'>', '>', '>'}, 0u, 0u, 0u, 0u, 0u, 0u};
-	static uint8_t seq = 0;
 
 	// [>,>,>,topic_ID,seq,payload_length,CRCHigh,CRCLow,payload_start, ... ,payload_end]
 	uint16_t crc = crc16((uint8_t *)&buffer[sizeof(header)], length);
 
 	header.topic_ID = topic_ID;
-	header.seq = seq++;
+	header.seq = _seq_number.fetch_add(1);
```
Comment (bkueng) [inline msg/templates/urtps/microRTPS_transport.cpp:230]: Why is this atomic?

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em [hunk 3]
```diff
@@ -149,10 +154,10 @@ static int launch_send_thread(pthread_t &sender_thread)
 {
     pthread_attr_t sender_thread_attr;
     pthread_attr_init(&sender_thread_attr);
-    pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(4000));
+    pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(6144));
     struct sched_param param;
     (void)pthread_attr_getschedparam(&sender_thread_attr, &param);
-    param.sched_priority = SCHED_PRIORITY_DEFAULT;
+    param.sched_priority = SCHED_PRIORITY_MAX - 5;
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em]: Sure. Most of your review comments are related with some trials before reaching the final solution. Reverted all.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em [hunk 4]
```diff
@@ -81,6 +81,9 @@ void* send(void *data);
 uint8_t last_remote_msg_seq = 0;
 uint8_t last_msg_seq = 0;
 
+pthread_mutex_t rcv_lock;
+pthread_mutex_t snd_lock;
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em]: Not needed for sure. Reverted.

File: msg/templates/urtps/microRTPS_agent.cpp.em [hunk 5]
```diff
@@ -276,6 +290,7 @@ int main(int argc, char** argv)
         ++loop;
         if (!receiving) start = std::chrono::steady_clock::now();
         // Publish messages received from UART
+        std::unique_lock<std::mutex> lk(rcv_mutex);
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/templates/urtps/microRTPS_agent.cpp.em]: True. Reverted.

File: src/modules/micrortps_bridge/micrortps_client/microRTPS_client_main.cpp [hunk 6]
```diff
@@ -195,8 +213,8 @@ int micrortps_client_main(int argc, char *argv[])
 
 		_rtps_task = px4_task_spawn_cmd("rtps",
 						SCHED_DEFAULT,
-						SCHED_PRIORITY_DEFAULT,
-						4096,
+						SCHED_PRIORITY_MAX - 5,
+						6144,
```
[the remark this answers is not in the document]
Comment (TSC21) [reply src/modules/micrortps_bridge/micrortps_client/microRTPS_client_main.cpp]: Yep this did the trick. Added. Thanks!

File: msg/templates/urtps/microRTPS_transport.h [hunk 7]
```diff
@@ -33,15 +33,23 @@
...
+#include <cstring>
 #include <poll.h>
+#if !defined (__PX4_NUTTX)
+#include <atomic>
+#else
+#include <px4_platform_common/atomic.h>
+#endif
 #include <termios.h>
 
+#define BUFFER_SIZE 2048
```
[the remark this answers is not in the document]
Comment (TSC21) [reply msg/templates/urtps/microRTPS_transport.h]: No need. Reverted.

File: msg/templates/urtps/microRTPS_transport.cpp (hunk 1 above)
Comment (TSC21) [reply msg/templates/urtps/microRTPS_transport.cpp:118]: This made more sense to me since we are using non-blocking reads, and overall the logic is kept the same.

File: msg/templates/urtps/microRTPS_transport.cpp (hunk 2 above)
Comment (TSC21) [reply msg/templates/urtps/microRTPS_transport.cpp]: No need. Reverted.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em [hunk 8]
```diff
@@ -149,7 +168,7 @@ static int launch_send_thread(pthread_t &sender_thread)
 {
     pthread_attr_t sender_thread_attr;
     pthread_attr_init(&sender_thread_attr);
-    pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(4000));
+    pthread_attr_setstacksize(&sender_thread_attr, PX4_STACK_ADJUSTED(4096));
```
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:171]: How much does it need now? (after the dynamic allocation)

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em [hunk 9]
```diff
@@ -89,11 +110,7 @@ void* send(void* /*unused*/)
     int loop = 0, read = 0;
     uint32_t length = 0;
     size_t header_length = 0;
-
-    /* subscribe to topics */
-@[for idx, topic in enumerate(send_topics)]@
-    uORB::Subscription @(topic)_sub{ORB_ID(@(topic))};
-@[end for]@
+    struct SendTopicsSubs subs;
```
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:113]: You need to allocate this dynamically, otherwise you continue to have the same problem.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em [hunk 10]
```diff
@@ -164,16 +183,10 @@ static int launch_send_thread(pthread_t &sender_thread)
...
     char data_buffer[BUFFER_SIZE] = {};
     int read = 0;
     uint8_t topic_ID = 255;
-
-    // Declare received topics
-@[for idx, topic in enumerate(recv_topics)]@
-    @(receive_base_types[idx])_s @(topic)_data;
-    uORB::Publication<@(receive_base_types[idx])_s> @(topic)_pub{ORB_ID(@(topic))};
-@[end for]@
+    struct RcvTopicsPubs pubs;
```
Comment (bkueng) [inline msg/templates/uorb_microcdr/microRTPS_client.cpp.em:189]: same as above

File: msg/templates/urtps/microRTPS_transport.cpp (hunk 1 above)
Comment (bkueng) [reply msg/templates/urtps/microRTPS_transport.cpp:118]: There's already a loop around the read call, so it's not required, but generally fine to keep it.
In which case is it executed more than once though? There's a `return len;` right at the bottom.

File: msg/templates/urtps/microRTPS_transport.cpp [hunk 11]
```diff
@@ -113,104 +113,101 @@ ssize_t Transport_node::read(uint8_t *topic_ID, char out_buffer[], size_t buffer
...
+			len = payload_len + header_size;
+		}
 
-	} else {
-		// copy message to outbuffer and set other return values
-		memmove(out_buffer, rx_buffer + msg_start_pos + header_size, payload_len);
-		*topic_ID = header->topic_ID;
-		len = payload_len + header_size;
-	}
+		// discard message from rx_buffer
```
Comment (bkueng) [inline msg/templates/urtps/microRTPS_transport.cpp:203]: This should not be done in case of a CRC error

File: msg/templates/urtps/microRTPS_transport.cpp (hunk 1 above)
Comment (TSC21) [reply msg/templates/urtps/microRTPS_transport.cpp]: Actually after your comment and a second thought on it, this while doesn't make sense. I just reverted it.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em (hunk 8 above)
Comment (TSC21) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:171]: I am actually making it consistent with the main task. But actually 4000 still works.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em (hunk 9 above)
Comment (TSC21) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:113]: Interesting enough without dynamically allocating it, it works without increasing the stack. But I'll do that.

File: msg/templates/uorb_microcdr/microRTPS_client.cpp.em (hunk 8 above)
Comment (bkueng) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em:173]: Consistency does not matter for that, make it as small as required (actual use + 350 bytes). Otherwise it's just wasting RAM.
The same for https://github.com/PX4/Firmware/blob/master/src/modules/micrortps_bridge/micrortps_client/microRTPS_client_main.cpp#L199.

Comment (TSC21) [reply msg/templates/uorb_microcdr/microRTPS_client.cpp.em]: Thanks for the feedback. Solved in https://github.com/PX4/Firmware/pull/15354/commits/730b1fe13494b411d5fb55cab285cb91d46cc71f.

**Comment by TSC21** [comment]
Thanks @bkueng!
